osquery/events/darwin/openbsm.cpp

```diff
@@ -16,10 +18,92 @@
 
 namespace osquery {
 
+const int kQLimit = 512;
+
 DECLARE_bool(disable_audit);
+DECLARE_bool(audit_allow_process_events);
+DECLARE_bool(audit_allow_sockets);
+DECLARE_bool(audit_allow_user_events);
+DECLARE_bool(audit_allow_fim_events);
 
 REGISTER(OpenBSMEventPublisher, "event_publisher", "openbsm");
 
+Status OpenBSMEventPublisher::configureAuditPipe(FILE* au_pipe) {
+  auto au_fd = fileno(au_pipe);
+  int pr_sel_mode = AUDITPIPE_PRESELECT_MODE_LOCAL;
+
+  audit_pipe_ = nullptr;
+
+  if (ioctl(au_fd, AUDITPIPE_SET_PRESELECT_MODE, &pr_sel_mode) == -1) {
+    LOG(WARNING) << "The auditpipe:ioctl AUDITPIPE_SET_PRESELECT_MODE failed";
+    fclose(au_pipe);
+    return Status(1, "Failed to set AUDITPIPE_SET_PRESELECT_MODE");
+  }
+
+  if (ioctl(au_fd, AUDITPIPE_SET_QLIMIT, &kQLimit) == -1) {
+    LOG(INFO) << "The auditpipe:ioctl AUDITPIPE_SET_QLIMIT failed";
+  }
+
+  struct au_class_ent* ace;
+  au_mask_t pr_flags = {0, 0};
+  std::vector<std::string> ev_classes;
+
+  if (true == FLAGS_audit_allow_process_events) {
+    // capture process events
+    ev_classes.push_back("pc");
+  }
+
+  if (true == FLAGS_audit_allow_sockets) {
+    // capture network events
+    ev_classes.push_back("nt");
+  }
+
+  if (true == FLAGS_audit_allow_user_events) {
+    // capture user (login, autherization etc...) events
+    ev_classes.push_back("lo");
+    ev_classes.push_back("aa");
+    ev_classes.push_back("ad");
+  }
+
+  if (true == FLAGS_audit_allow_fim_events) {
+    // capture file events
+    ev_classes.push_back("fc");
+    ev_classes.push_back("fd");
+    ev_classes.push_back("fw");
+    ev_classes.push_back("fr");
+    ev_classes.push_back("fa");
+    ev_classes.push_back("fm");
+  }
+
+  while ((ace = getauclassent()) != nullptr) {
+    for (auto& cl : ev_classes) {
+      if (cl == ace->ac_name) {
+        ADD_TO_MASK(&pr_flags, ace->ac_class, AU_PRS_BOTH);
+        break;
+      }
+    }
+  }
+  endauclass();
+
+  au_mask_t na_pr_flags = pr_flags;
+
+  if (ioctl(au_fd, AUDITPIPE_SET_PRESELECT_FLAGS, &pr_flags) == -1) {
+    LOG(WARNING) << "The auditpipe:ioctl AUDITPIPE_SET_PRESELECT_FLAGS failed";
+    fclose(au_pipe);
+    return Status(1, "Failed to set AUDITPIPE_SET_PRESELECT_FLAGS");
+  }
+
+  if (ioctl(au_fd, AUDITPIPE_SET_PRESELECT_NAFLAGS, &na_pr_flags) == -1) {
+    LOG(WARNING)
+        << "The auditpipe:ioctl AUDITPIPE_SET_PRESELECT_NAFLAGS failed";
+    fclose(au_pipe);
+    return Status(1, "Failed to set AUDITPIPE_SET_PRESELECT_NAFLAGS");
+  }
+
+  audit_pipe_ = au_pipe;
+  return Status(0);
```

Comment:
nitpick, for all of the `Status` calls, do you mind using the new `Status::success()` and `Status::failure("failure note")` APIs?